[382] ClearCheckSum support for handling changed migrations

gonzalad opened the pull request:
This is a very rough attempt to implement clearCheckSum support in liquigraph 4.0.

One use case of this feature is to handle changing existing migrations in scenarii such as neo4j db upgrades (when the old migrations use a no longer supported syntax).

The initial support works like this:

. need to rewritte the faulty migrations using the new syntax.
. execute Liquigraph CLI qui the clear-checksum flag
```
./liquigraph.sh --changelog "${PROJECT_HOME}/changelog.xml" \
    --username neo4j \
    --password \
    --graph-db-uri jdbc:neo4j:bolt://localhost:7687/ \
    --clear-checksum
``` 
. then run liquigraph (as always)
 In this step, liquigraph will detect changeset without checksum and will recompute the checksum (without executing the changeset)

I think there's quite some work to do on refactoring the PR and, once it's done, add more tests.  

I'll gladly accept any guidance on the areas to improve !!!

Thanks,
Adrian

Commit:
feature/clearCheckSums: initial support (wip)

TODO:

- refactor
- UT

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/api/MigrationRunner.java:
```diff
@@ -90,6 +98,22 @@ public void runMigrations(Configuration configuration) {
         return persistedChangesets;
     }
 
+    private void replaceChecksum(Configuration configuration,
```
I think this checksum replacement would be better in the `ChangelogGraphReader` class
It may look counter-intuitive to perform writes in a `XxxReader` class, but it already does to perform internal migrations.
In a way, `ChangelogGraphReader` output invariant is that it always returns a checksummed sequence of persisted migrations.

gonzalad replied:
> I think this checksum replacement would be better in the ChangelogGraphReader

I agree 100%.

WDYT about creating a ChangelogGraphReader interface that will be implemented both by the current ChangelogGraphReader and ReplaceChecksumGraphReader ?

ReplaceChecksumGraphReader will the delegate to ChangelogGraphReader to retrieve the changesets from db, and will check if there are some with null checksum, if that's the case it will update the corresponding changesets in db.

So to resume, we'll have something like

```

    private ChangelogReader changelogReader(Collection<Changeset> declaredChangesets){
      return new ReplaceChecksumGraphReader(declaredChangesets, changelogReader);
    }

    private Collection<Changeset> readPersistedChangesets(Collection<Changeset> declaredChangesets, Connection writeConnection) {
        ChangelogGraphReader reader = changelogReader(declaredChangesets);
        Collection<Changeset> persistedChangesets = reader.read(writeConnection);
       ... as before...
    }
```

gonzalad replied:
done in commit 9754dceb09ff62ed1b67b7522c0dd7b681e48feb

fbiville replied:
I understand better now what you intended to do, thanks for taking the time to lay out your ideas.

I'm not 100% convinced decorator is a good fit here, as I do not need to compose different changelog reading strategies. Indeed, changesets without checksums should _always_ have their checksums re-computed, it may be easier to implement that part directly in the existing `ChangelogGraphReader`.

gonzalad replied:
done in commit 167e6de

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/api/Runner.java:
```diff
@@ -0,0 +1,34 @@
+     *
+     * @param configuration contains all the configuration data needed for running this migration
+     */
+    void runMigrations(Configuration configuration);
```
probably better to generalize the name to just `run`?
In the case of checksum, no migrations are run at all.

gonzalad replied:
done in commit 9754dceb09ff62ed1b67b7522c0dd7b681e48feb

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/api/Runner.java:
```diff
@@ -0,0 +1,34 @@
+public interface Runner {
+
+    /**
+     * Execute the migrations specified by configuration parameter
```
This needs to be generalized (or removed) as this description is not true for all implementations.

gonzalad replied:
done in commit 9754dceb09ff62ed1b67b7522c0dd7b681e48feb

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/configuration/ClearChecksumMode.java:
```diff
@@ -0,0 +1,27 @@
+ * On next update changesets that have already been deployed will have their checksums
+ * recomputed, and changesets that have not been deployed will be deployed.
+ */
+public enum ClearChecksumMode implements ExecutionMode {
```
I'm not entirely sure this is a good fit.
I don't think `clear checksum` is a mode of the regular Liquigraph execution, but rather a new action.
Instead, `clearChecksums()` should be a new API on the main `Liquigraph` object

gonzalad replied:
Right, I'll add the Liquigraph.clearChecksums() and I'll remove this mode and revert the Runner interface (don't think it will be needed anymore)

gonzalad replied:
done in commit 9754dceb09ff62ed1b67b7522c0dd7b681e48feb

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/io/ReplaceChecksumWriter.java:
```diff
@@ -0,0 +1,77 @@
+
+    private static final String REPLACE_CHECKSUM_UPDATE =
+        "MATCH (changeset:__LiquigraphChangeset {id: ?, author: ?}) " +
+            "SET  changeset.checksum = ?";
```
it should happen for changesets that currently do not have a changeset (it's probably the case as you currently filter the changesets beforehand) but since this query is likely to move to `ChangelogGraphReader`, this query will need to be updated

gonzalad replied:
In commit 9754dce: The writer is used as is and called by ReplaceChecksumGraphReader which filters the changesets to update before calling the writer.

fbiville reviewed (changes requested):
This is not a thorough review. 
But this going in the right direction, I think just a couple of adjustments are needed.

Commit:
feature/clearCheckSums: feedback from initial PR review

. CLI: new action parameter run/dry-run/clear-checksum
. ReplaceChecksum functionnality was moved to reader
. Runner API was removed
. Liquigraph: clearChecksums() method was added

gonzalad commented on liquigraph-cli/src/main/java/org/liquigraph/cli/LiquigraphCli.java:
```diff
@@ -105,6 +106,13 @@
     )
     private String dryRunOutputDirectory;
 
+    @Parameter(
```
done in commit 9754dceb09ff62ed1b67b7522c0dd7b681e48feb

gonzalad commented:
Thanks for your initial feedback, it's very constructive. 

I pushed a new commit including fixes for all your inital feedback and completing test coverage.

Don't hesitate if you have any further feedback !

This commit is not tested throughly (it's more to gather more feedback about the design).

In parallel, I'll test this commit to migrate my current app to neo4j 4.x.

Thanks,
Adrian

gonzalad commented:
I just migrated our app with this branch, it works !

I tested:

- clearing checksum with CLI
  <check in db that checksum attribute is not present in the changeset>
- running the app with liquigraph enabled (spring boot app)
  <check in db that checksum attribute is now present in the changeset>
  <check that migrations were not executed>
- clear checksum with maven plugin

gonzalad commented on liquigraph-core/src/main/java/org/liquigraph/core/model/Changeset.java:
```diff
@@ -147,8 +146,7 @@ public boolean equals(Object obj) {
         final Changeset other = (Changeset) obj;
         return Objects.equals(this.id, other.id) &&
-            Objects.equals(this.author, other.author) &&
-            Objects.equals(this.checksum, other.checksum);
```
Please note that I changed changeset equals/hashcode contract: it doesn't take checksum into account anymore.

I didn't see any impact, but I think it's important enough to write this comment just in case I missed something...

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/io/ReplaceChecksumGraphReader.java:
```diff
@@ -0,0 +1,86 @@
+import static java.util.stream.Collectors.toList;
+import static org.liquigraph.core.model.predicates.ChangesetChecksumIsEmpty.CHECKSUM_IS_EMPTY;
+
+public class ReplaceChecksumGraphReader implements ChangelogGraphReader {
```
I don't think replacing checksums is related to read changelogs from a graph database (as `ChangelogGraphReader` suggests)
the operation of replacing checksums should be done when reading changesets in the original `ChangelogGraphReader` implementation, in my opinion

gonzalad replied:
Hi @fbiville,

Sorry, but I don't understand your suggestion.

ReplaceChecksumGraphReader is just a decorator delegating read functionnality to it delegate.
It gets called when MigrationRunner needs to read a persisted changelog:
MigrationRunner -> ReplaceChecksumGraphReader -> ChangelogGraphReaderImpl
and avoids to clutter ChangelogGraphReaderImpl with code related to updating checksums.

Do you prefer me to:

1. move the code from ReplaceChecksumGraphReader inside of ChangelogGraphReaderImpl ?
   i.e. something like calling a private method replaceEmptyChecksums that I define in ChangelogGraphReaderImpl from the ChangelogGraphReaderImpl#read ?
2. merge ReplaceChecksumGraphReader and ReplaceChecksumWriter classes and call this new class from ChangelogGraphReaderImpl#read method ? (similar to 1 in fact).
3. other ?

Thanks,
Adrian

fbiville replied:
Hi, apologies for my unclear comment.
Indeed, I think option 1 would make more sense.

gonzalad replied:
done in commit 167e6de

fbiville commented:
By the way, do not worry about the listed conflicts, I pushed a big change to the tests while you were working on this PR.
I will take care of rebasing your work and adapt your tests.

fbiville commented on liquigraph-cli/src/main/java/org/liquigraph/cli/LiquigraphCli.java:
```diff
@@ -105,11 +108,19 @@
+            "dry-run (default if --dry-run-output-directory is set), " +
+            "clear-checksum"
+    )
+    private List<String> parameters = new ArrayList<>();
```
If I understand the JCommander docs correctly, you can have a main parameter (i.e. without `names` attribute like you did above) that can be defined as of type`String`.
I assume JCommander will parse the action correctly and the custom parsing you do in the `action` method will not be necessary anymore (or at least drastically reduced).

fbiville replied:
You can even implement a [custom converter](https://jcommander.org/#_custom_types_converters_and_splitters) so that the field is of type `Action`.

gonzalad replied:
> If I understand the JCommander docs correctly, you can have a main parameter

The issue is that JCommander sends all unknown parameters into parameters

i.e. when I use the following parameters

```
--changelog "../../../../cof-server-v2/infrastructure/neo4j-adapter/neo4j-adapter-migration/src/main/resources/db/liquigraph/changelog.xml" --username neo4j --password "toto" --graph-db-uri jdbc:neo4j:bolt://localhost:7688/ --unkown dry-run
```

I will have parameters (size 2) containing dry-run and --unkown.

What I can do is just check that parameters size is less or equals to 1 (and remove the extra filtering on --<option> I'm doing).

gonzalad replied:
Also, if I add arity = 1 to parameters, I'll have the following error message

```
com.beust.jcommander.ParameterException: There should be exactly 1 main parameters but 2 were found
```

Which is not very helpfull for the user since JCommander doesn't tell the user the name of the parameter in cause.

That's why I added code to check the length of parameters array

gonzalad replied:
Ok, I tested with the converter approach.

I'll create a converter that throws a ParameterException if the argument is not a valid action.

i.e. 
```
Exception in thread "main" com.beust.jcommander.ParameterException: Parameter action '--unkown' is invalid, must be one of DRY_RUN, RUN, CLEAR_CHECKSUM
```

gonzalad replied:
done in commit eee1086c

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/api/ClearChecksumRunner.java:
```diff
@@ -0,0 +1,41 @@
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClearChecksumRunner.class);
+
+    public void run(Configuration configuration) {
+        Supplier<Connection> connectionSupplier = new ConnectionSupplier(new GraphJdbcConnector(configuration));
```
You can use `GraphJdbcConnector` and call `GraphJdbcConnector#connect` directly, you do not need to go through `ConnectionSupplier`

gonzalad replied:
done in commit eee1086c

I removed the class ConnectionSupplier and restored the previous codebase in its original state

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/api/Liquigraph.java:
```diff
@@ -70,8 +69,8 @@ private static ChangelogDiffMaker changelogDiffMaker() {
 
-    private static ChangelogGraphReader changelogGraphReader() {
-        return new ChangelogGraphReader();
+    private static ChangelogGraphReaderImpl changelogGraphReader() {
```
per our previous discussion (about implementing `clearChecksums` as a private method in `ChangelogGraphReader`, this change won't be necessary

gonzalad replied:
done in commit eee1086c

But I needed to change ChangelogGraphReader with a Function<Collection<Changeset>, ChangelogGraphReader> since the graphReader now needs the declaredChangeSet.

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/configuration/ConfigurationBuilder.java:
```diff
@@ -165,7 +164,19 @@ public ConfigurationBuilder withRunMode() {
+     * @param executionMode the execution mode to run
+     * @return itself for chaining purposes
+     */
+    public ConfigurationBuilder withExecutionMode(ExecutionMode executionMode) {
```
can probably become private (as there are no new execution modes anymore)

gonzalad replied:
done in commit eee1086

I reverted this change

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/io/ChangelogGraphReader.java:
```diff
@@ -16,89 +16,14 @@
+/**
+ * Interface to retrieve the persisted changelog from neo4j db.
+ */
+public interface ChangelogGraphReader {
```
per our previous discussion (about implementing clearChecksums as a private method in ChangelogGraphReader, this change won't be necessary

gonzalad replied:
done in 167e6deda602faa5e88b77fc4d14e246a8b4fcf5

fbiville commented on liquigraph-core/src/main/java/org/liquigraph/core/io/ConnectionSupplier.java:
```diff
@@ -0,0 +1,33 @@
+import java.sql.Connection;
+import java.util.function.Supplier;
+
+public class ConnectionSupplier implements Supplier<Connection> {
```
optional: you can even rename it to `LockedConnectionSupplier` to convey the semantics of `GraphJdbcConnector` (which should probably be renamed to `LiquigraphJdbcConnector` as the locking part is specific to Liquigraph)

Note: this change may not be necessary as `ClearChecksumRunner` does not need the supplier (see comment above)

gonzalad replied:
After reading one of your previous comments, I wonder why we use a ConnectionSupplier instead of calling GraphJdbcConnector#connect directly here.

I'll revert this change to keep the original codebase here.

gonzalad replied:
done in eee1086cc909db4c63be8da1800206232d13b6c7

fbiville commented on liquigraph-core/src/test/java/org/liquigraph/core/io/ChangelogGraphReaderTestSuite.java:
```diff
@@ -39,7 +39,7 @@
     }
 
-    private ChangelogGraphReader reader = new ChangelogGraphReader();
+    private ChangelogGraphReaderImpl reader = new ChangelogGraphReaderImpl();
```
per our previous discussion (about implementing clearChecksums as a private method in ChangelogGraphReader, this change won't be necessary

gonzalad replied:
done in 167e6deda602faa5e88b77fc4d14e246a8b4fcf5

fbiville commented on liquigraph-core/src/test/resources/changelog/changelog-with-2-nodes.xml:
```diff
@@ -0,0 +1,26 @@
+-->
+<changelog>
+    <changeset id="insert-fbiville" author="fbiville">
+        <query><![CDATA[CREATE (:Human {name:'fbiville'})]]></query>
```
optional: you can remove `<![CDATA[` and `]]>` given there is no `>` or `<` in the query contents (that usually happens when matching relationships)

gonzalad replied:
Most of the existing changelogs in this folder use CDATA in the same condition (no < and > in the query).

The best imo will be to update all the changelogs to be consistent. 

In the meantime, I'll leave this as-is if it's ok for you.

fbiville reviewed (changes requested):
Getting there, thanks for your patience and efforts :)

Commit:
feature/clearCheckSums: merge ReplaceChecksumReader in main reader

Merge ReplaceChecksumGraphReader into ChangelogGraphReader

Commit:
feature/clearCheckSums: feedback from PR review

- use JCommander converter
- use GraphJdbcConnector instead of ConnectionSupplier
- fix builder usage (in order to comply with a future non-mutable builder)
- removed withExecutionMode method (not necessary anymore)

gonzalad commented on liquigraph-cli/src/main/java/org/liquigraph/cli/LiquigraphCli.java:
```diff
@@ -131,17 +142,18 @@ public static void main(String[] args) {
+            new Liquigraph().clearChecksums(builder.build());
+        } else {
+            if (action == Action.DRY_RUN) {
+                builder.withDryRunMode(Paths.get(cli.dryRunOutputDirectory));
```
done in commit eee1086c

gonzalad commented:
Hello @fbiville,

One more question on this PR.

The current behaviour of clear-checksum is to clear the checksum from all the persisted changesets.

WDYT if we clear only the checksum for the persisted changesets which have a matching declared changeset.

This would handle the scenario of 2 applications sharing the same neo4j database.

i.e. app1 needs to clear the checksum from its on migrations without affecting the changesets from app2.

The cons is that for that to work we would need to have access to the declaredChangeset when executing the clear-checksum command.

WDYT ?

Thanks,
Adrian

fbiville commented:
Hey @gonzalad, another option would to introduce a new option that could act as a changeset selector, which could be:

-  '*' (all persisted changesets)
- 'some-changeset-id' to match the changeset that has this ID
- 'some-id-1,some-id-2,some-id-3' to match the changesets that have one of these IDs

'*' could maybe be the default or not (for reasons you explained above). 
It may be better to fail fast if the selector is not set.

What do you think?

gonzalad commented:
> What do you think?

The issue with that method is that the user needs to know all the changeset-its he needs to clear (i.e. if my app1 has 1000 changesets, he needs to know and list them all).

Using the local changeset would avoid this issue (but of course it would require having those changesets locally :) ).

Additional solutions could be:

1. Adding more options to the CLI: We could also imagine having some kind of regex (for computing the changesetid to clear), have some more options exclude-change-ids or exclude-change-ids-regex...
1. Adding an id in changelog
  We could also have different changelogs objects in db (each one corresponding to a given declared changelog.xml) - but this would need to introduce id in the changelog.
1. Use different neo4j liquigraph collections 
  We could parameterize the name of the neo4j collections used by liquigraph to separate the collections from app1 and app2 (i.e something like databaseChangeLogTableName from liquibase https://docs.liquibase.com/tools-integrations/ant/home.html ? don't know if using that option is enough with liquibase).

Perhaps we should tackle this issue in a separate PR ? (to take the time to think a bit more about it) 

My question was perhaps a bit premature :)

fbiville commented:
Actually, I've been thinking about the item 2 and 3 for a few days already.

While it is not strictly related, the fact that Neo4j 4 (Enterprise Edition) is now multi-tenant, we could imagine Liquigraph history graph be stored in a different database than the one where the data to migrate is.

And even without multi-tenancy, it could make sense to have different Liquigraph history graphs for different apps in the same DB as you suggested.

I see these two as two new, separate issues to tackle.

fbiville commented:
Let's imagine the two above issues are solved now and let's see what's needed for clearing checksums. I see two things:

- the history graph ID, or changelog ID as you called it (as commented above, there may be several history graphs per database)
- the ID(s) of the changesets to clear

- If one needs to clear all changesets of a given history graph, then '*' could be a value that represents all IDs.
- If one needs to clear all changesets but a couple of ones, we could imagine having two parameters, includes and excludes, where `includes = '*'` and `excludes = 'changeset-1,changeset-2'`.
- If one needs to clear a couple of changesets, then it would be `includes = 'changeset-1,changeset-2'`.

Then, there is the muddy area where one needs to clear many changesets, but that's only a fraction of the whole. Regex could help so long as the IDs share a similar structure.

I would suggest for now to start with a simple version: let's explicit list the changeset IDs.
We can evolve this later with wildcard, regex and exclusion.
And even later, maybe introduce a new command that gives the ordered list of changeset IDs of a given history graph stored in a specific database.

What do you think?

gonzalad commented:
> I would suggest for now to start with a simple version: let's explicit list the changeset IDs.
> What do you think?

Fine for me, I'll implement that as an extra option `--include <changeset-list>` in Liquigraph CLI.
With changeset-list syntax being: <changeset-1>,<changeset-2>

Right ?

fbiville commented:
> Fine for me, I'll implement that as an extra option `--include <changeset-list>` in Liquigraph CLI.

Note this should be an option for the Maven plugin and the Spring Boot autoconfiguration as well.

> With changeset-list syntax being: ,
> 
> Right ?

That's correct, canonical syntax should be `item1,item2,item3` but we can tolerate `item1, item2, item3`.

Commit:
feature/clearCheckSums: add include option

Included option is added to both LiquiGraphCli and Maven mojo.

The included option must be a list of changeset ids.

When this option is set and when we execute the clear-checksum
command, then only the checksum of the given changesets are cleared.

Otherwise all checksum are cleared.

gonzalad commented:
Hi @fbiville,

I implemented the include option in both Maven and CLI.
Not in spring boot. 

Reasons: 

* I didn't implemented clear-checksum action in spring boot in the previous commits
* I don't see the use case running clear-checksum inside a spring boot app (the classic use case would be to run the migrations from a spring boot app, and if need be to use a script - outside of the app - to run the clear checksum => hence CLI or Maven job or direct liquigraph API usage)

About the option name, I used 'include' in the CLI and 'includes' in the maven mojo, perhaps I should use only a single name, in this case which one ?

Thanks,
Adrian

fbiville commented:
@gonzalad makes sense for Spring Boot 👍

fbiville commented:
>  I used 'include' in the CLI and 'includes' in the maven mojo

I'd vote for `include`. No strong preference, though.

fbiville commented:
@gonzalad I have an idea for a CLI refactoring that will make it easier to add clear-checksum (I invited to a private repo showing what it could look like).

May I suggest I take it from there? In any, you'll properly be credited the changes.

I don't want you to go through 2 painful rebases (first being the already-merged testcontainers migration, second being the future CLI refactoring).

Commit:
feature/clearChecksums: use include parameter in mvn and CLI

Use same name for include parameter of clear-checksum command

gonzalad commented:
> I'd vote for `include`. No strong preference, though.

Done :)

gonzalad commented:
> May I suggest I take it from there? In any, you'll properly be credited the changes.

Gladly, no problem at all ! 

I looked at your git repo, it's definitely cleaner.

On a side note, perhaps a tiny bit of cleanup is required on the Maven commands too (the base mojo relies on the include parameter but this parameter should be private to ClearChecksumMojo).

Thanks @fbiville !

fbiville commented:
> On a side note, perhaps a tiny bit of cleanup is required on the Maven commands too (the base mojo relies on the include parameter but this parameter should be private to ClearChecksumMojo).

Absolutely :)

fbiville commented:
CLI has been refactored.
Maven plugin on the way: https://github.com/liquigraph/liquigraph/pull/406 (not as pretty as I would have wanted but that will make our work here easier).
Once the above PR merged (feel free to have a look), I will pick up your changes.

gonzalad commented:
Hi @fbiville, what's the future of this feature ?
I see the PR was closed yesterday without being merged and without explanation.
Thanks !

fbiville commented:
Hi, sorry, it's the rename of the master branch that caused this.
I'm still working on backporting your changes, it's taking a bit more time than I expected.
